### ISSUE TITLE
Expand All/Collapse ALL Problem with Tree View

### ISSUE THREAD

**Issue by dereklin** [angular-ui/ng-grid#3399]
This issue happens both in the tutorial: http://ui-grid.info/docs/#/tutorial/215_treeView

Or in my plunker: http://plnkr.co/edit/MwhZMkAeioxceQtbChN1?p=preview

When you manually expand individual sub-trees, the Master Tree still shows the + sign.  Instead, it should show the - sign, since all sub-trees have been expanded.

Vice Versa for Collapse ALL.

**Comment by PaulL1** [comment]
Yes, there isn't a check as to whether every tree node is expanded, it drives off whether you clicked expand all or programatically called expand all.

I'm not keen to write a check as to whether every node is expanded - that would get expensive.  But I would agree that (I think like grouping) if you have expanded all and you collapse something, then expandAll should toggle back to false.

**Comment by dereklin** [comment]
When I expand all sub-trees, the master expand/collapse still shows +.  It should show - because all sub-trees are already expanded.  I might have missed it somewhere in your message.  Did you say fixing this will be too expensive and therefore, it won't be fixed or changed?

**Comment by PaulL1** [comment]
Yes.  It would mean iterating over the entire tree on every digest cycle to see whether or not every cell is expanded.  If we have a very large data set this will be expensive.  Most applications don't do this - you only go to a - if someone explicitly clicks the button (or, in our case, calls expand all programatically).

**Comment by dereklin** [comment]
Is there an option to hide the master expand/collapse button?

And would it be feasible to keep a tally for number of expanded/collapsed sub-trees on each expand/collapse action, and then just compare the tally with the total number of sub-trees to show/hide the master expand/collapse button?

For now, I think I can just use css to hide the master expand/collapse button.

**Comment by PaulL1** [comment]
There is not currently an option to hide the button.

**Comment by dereklin** [comment]
@PaulL1 This is actually fixed!  I thought you said it would be too expensive to fix... Anyways, it's great.  But with the new changes, my decorator is broken, that's a different story...

**Comment by PaulL1** [comment]
Yeah, I did it in a different way that didn't require polling/binding, so performance impact is less.

**Comment by tringuyenminh92** [comment]
how to know all rows are expanded after calling $scope.gridApi.treeBase.expandAllRows() method?
Because i need to call    $scope.gridApi.core.scrollTo after calling expandAllRows. but it shown error message: 
Error: grid.treeBase.tree.forEach is not a function
service.expandAllRows@http://localhost:51358/Scripts/angular/ui-grid.js:24508:9
service.initializeGrid/publicApi.methods.treeBase.expandAllRows@http://localhost:51358/Scripts/angular/ui-grid.js:24090:17

But if i let it loaded data and expanded. After that some seconds, i click button to call scrollTo method.
it will work.

**Comment by mattaGIT** [comment]
None of the above worked for me for all of my grid use cases. 


            $scope.gridApi.grid.registerDataChangeCallback(function() {
                if($scope.gridApi.grid.treeBase.tree instanceof Array){
                    $scope.gridApi.treeBase.expandAllRows();
                }
                
            });


The following works in every case I have tested. DataChangeCallback is called twice (for some unknown reason) on initial page load. The first time, gridApi.grid.treeBase.tree is an object which causes the issue with gridApi.grid.treeBase.tree.forEach above:

**Comment by ramseyfeng** [comment]
@PaulL1, has this issue been fixed already?
I think you only fixed one way: unset expandAll icon when collapse some row, correct?
What about the other way? like you manually expand every rows one by one, will the the expandAll icon be synchronized?
 I saw previously you have answered it would be expensive and you won't like to fix it.
But from your last reply, you said it has been fixed it, did you mean you have fixed this way as well? Or you only fix one of the scenario: "**unset expandAll icon when collapse some row**"
